From e1420eca7374cd8f583e9d774c890645a205aaee
From: Bernhard Reutner-Fischer
Date: Tue, 14 Jun 2011 14:31:00 +0000
Subject: resolv: try next server on SERVFAIL

in bug 3637 Andrey Kovalev aka pxe.ru writes:
getaddrinfo does NOT add domain to query when receive SERVFAIL

RFC1035 7.2 suggests that
   - If a resolver gets a server error or other bizarre response
     from a name server, it should remove it from SLIST, and may
     wish to schedule an immediate transmission to the next
     candidate server address.

So let's try the next server upon SERVFAIL even if it's not strictly
required.

Signed-off-by: Bernhard Reutner-Fischer <rep.dot.nop@gmail.com>
---

From: Michel Stam <michel at reverze.net>
Date: Thu, 27 Feb 2014 21:42:38 +0100
Subject: [PATCH] resolv: try next server on SERVFAIL

Commit e1420eca7374cd8f583e9d774c890645a205aaee fixed a bug where a
response code should mean the next server is tried. However, it tries
only the next search domain, and never skips to the next server. This
fix makes sure we try the next server on SERVFAIL.

Ref.: http://lists.uclibc.org/pipermail/uclibc/2014-March/048268.html

Signed-off-by: Michel Stam <michel at reverze.net>
---

 libc/inet/resolv.c |   10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/libc/inet/resolv.c b/libc/inet/resolv.c
--- a/libc/inet/resolv.c
+++ b/libc/inet/resolv.c
@@ -1494,9 +1494,11 @@ int __dns_lookup(const char *name,
 				}
 				/* no more search domains to try */
 			}
-			/* dont loop, this is "no such host" situation */
-			h_errno = HOST_NOT_FOUND;
-			goto fail1;
+			if (h.rcode != SERVFAIL) {
+				/* dont loop, this is "no such host" situation */
+				h_errno = HOST_NOT_FOUND;
+				goto fail1;
+			}
 		}
 		/* Insert other non-fatal errors here, which do not warrant
 		 * switching to next nameserver */
-- 
